File: miniapp/src/services/api/product.ts
// 产品相关 API
import { apiClient } from './client'
import type { Product, PaginatedResponse, ProductQueryParams } from '@/types'

export class ProductApi {
  // 转换后端产品数据为前端格式
  transformProduct(backendProduct: any): Product {
    // 解析 countries JSON 字符串
    let coverageAreas: string[] = []
    let countryCode = ''
    let country = ''

    try {
      if (backendProduct.countries) {
        const countriesData = JSON.parse(backendProduct.countries)
        if (Array.isArray(countriesData) && countriesData.length > 0) {
          coverageAreas = countriesData.map((c: any) => c.cn || c.name || '')
          countryCode = countriesData[0].code || ''
          country = countriesData[0].cn || countriesData[0].name || ''
        }
      }
    } catch (e) {
      console.warn('解析国家数据失败:', e)
    }

    // 解析 features JSON 字符串
    let features: string[] = []
    try {
      if (backendProduct.features) {
        features = JSON.parse(backendProduct.features)
      }
    } catch (e) {
      console.warn('解析特性数据失败:', e)
    }

    // 格式化数据量
    const dataAmount = backendProduct.data_size >= 1024
      ? `${(backendProduct.data_size / 1024).toFixed(1)}GB`
      : `${backendProduct.data_size}MB`

    return {
      id: String(backendProduct.id),
      name: backendProduct.name || '',
      description: backendProduct.description || '',
      price: backendProduct.price || 0,
      originalPrice: backendProduct.retail_price || undefined,
      currency: 'USD',
      region: backendProduct.type || 'global',
      country: country,
      countryCode: countryCode,
      dataAmount: dataAmount,
      validDays: backendProduct.valid_days || 0,
      coverage: backendProduct.name || '',
      coverageAreas: coverageAreas,
      features: features,
      icon: backendProduct.image || undefined,
      isActive: backendProduct.status === 'active',
      isPopular: backendProduct.is_hot || false,
      createdAt: backendProduct.created_at || new Date().toISOString(),
      updatedAt: backendProduct.updated_at || new Date().toISOString()
    }
  }

  // 转换后端响应为前端格式
  transformProductResponse(backendData: any): PaginatedResponse<Product> {
    const { products = [], total = 0, limit = 20, offset = 0 } = backendData
    const page = Math.floor(offset / limit) + 1
    const totalPages = Math.ceil(total / limit)

    // 转换每个产品数据
    const transformedProducts = products.map((p: any) => this.transformProduct(p))

    return {
      items: transformedProducts,
      total,
      page,
      pageSize: limit,
      totalPages,
      hasNext: offset + limit < total,
      hasPrev: offset > 0
    }
  }

  // 获取产品列表
  async getProducts(params?: ProductQueryParams): Promise<PaginatedResponse<Product>> {
    // 转换前端参数为后端格式
    const backendParams: any = {}
    if (params?.type) backendParams.type = params.type
    if (params?.country) backendParams.country = params.country
    if (params?.search) backendParams.search = params.search
    if (params?.page) {
      const pageSize = params.pageSize || 20
      backendParams.limit = pageSize
      backendParams.offset = (params.page - 1) * pageSize
    } else {
      backendParams.limit = params?.pageSize || 20
      backendParams.offset = params?.offset || 0
    }

    const data = await apiClient.get('/miniapp/products', backendParams)
    return this.transformProductResponse(data)
  }

  // 获取产品详情
  async getProduct(id: string): Promise<Product> {
    const data = await apiClient.get(`/miniapp/products/${id}`)
    return this.transformProduct(data)
  }

  // 根据类型获取产品
  async getProductsByType(type: string, params?: Omit<ProductQueryParams, 'type'>): Promise<PaginatedResponse<Product>> {
    return this.getProducts({ ...params, type })
  }

  // 根据国家获取产品
  async getProductsByCountry(country: string, params?: Omit<ProductQueryParams, 'country'>): Promise<PaginatedResponse<Product>> {
    return this.getProducts({ ...params, country })
  }

  // 搜索产品
  async searchProducts(query: string, params?: ProductQueryParams): Promise<PaginatedResponse<Product>> {
    return this.getProducts({ ...params, search: query })
  }
}

// 创建产品 API 实例
export const productApi = new ProductApi()

File: backend/storage/models/product.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Product 产品模型
type Product struct {
	ID             int            `gorm:"primaryKey;autoIncrement" json:"id"`
	ThirdPartyID   string         `gorm:"uniqueIndex;size:100" json:"third_party_id"` // 第三方产品ID
	Name           string         `gorm:"size:200;not null" json:"name"`
	NameEn         string         `gorm:"size:200" json:"name_en"`
	Description    string         `gorm:"type:text" json:"description"`
	DescriptionEn  string         `gorm:"type:text" json:"description_en"`
	Type           string         `gorm:"size:20;index" json:"type"`  // local, regional, global
	Countries      string         `gorm:"type:text" json:"countries"` // JSON 格式存储国家列表
	DataSize       int            `gorm:"not null" json:"data_size"`  // MB
	ValidDays      int            `gorm:"not null" json:"valid_days"`
	Features       string         `gorm:"type:text" json:"features"` // JSON 格式存储特性列表
	Image          string         `gorm:"size:500" json:"image"`
	Price          float64        `gorm:"type:decimal(10,2)" json:"price"`
	CostPrice      float64        `gorm:"type:decimal(10,2)" json:"cost_price"`
	RetailPrice    float64        `gorm:"type:decimal(10,2)" json:"retail_price"`
	AgentPrice     float64        `gorm:"type:decimal(10,2)" json:"agent_price"`
	PlatformProfit float64        `gorm:"type:decimal(10,2)" json:"platform_profit"`
	IsHot          bool           `gorm:"default:false" json:"is_hot"`
	IsRecommend    bool           `gorm:"default:false" json:"is_recommend"`
	SortOrder      int            `gorm:"default:0" json:"sort_order"`
	Status         string         `gorm:"size:20;default:'active'" json:"status"` // active, inactive
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// TableName 指定表名
func (Product) TableName() string {
	return "products"
}
